src/core/client.ts

```diff
@@ -252,21 +252,28 @@ class ApiClient<R extends RawApi> {
 
             const res = await new Promise<ApiResponse<ApiCallResult<M, R>>>(
                 (resolve, reject) => {
-                    function onStreamError(err: unknown) {
+                    const onHttpError = toHttpError(method, sensLogs, reject);
+                    function onOtherError(err: unknown) {
                         abort();
```

Comment:
won't calling abort() cause `reject` to be called through `onHttpError`?

repeated calls to `reject` are ignored, but this is a code smell usually, with Node.js you can even track something like that: https://nodejs.org/api/process.html#event-multipleresolves

Reply:
True, it could cause several rejections because `fetch` will eventually also throw when the request is cancelled. We could introduce a flag `rejected` that prevents this, but I don't like that because I'm already unhappy with how hacky this gets. Do you have any suggestions?

Reply:
I guess we can simply remove this `reject(err)` after abort? Aborting request should immediately throw and won't slow anything down or something

Reply:
We cannot remove it because that would change which error is thrown. It will no longer be clear the the request timed out. Instead, it only looks like the request was cancelled, which is confusing.

Reply:
Update: we can now do it! The web spec was changed in https://github.com/whatwg/dom/pull/1027 and Deno 1.16 has already shipped support. I will update this PR accordingly as soon as I know how to achieve this on Node.

Reply:
Well, it does not look like we're going to get support for this anytime soon. The `abort-controller` package that is used under Node.js had its last publish 3 years ago, and the repo is pretty dead.

We now have several options.
1. Implement a hacky version with flags and checks to prevent repeated rejections.
2. Implement a new abstraction that works similarly to `deferred` from https://deno.land/std and that catches and suppressed subsequent rejections when already rejected.
3. Ignore the problem and simply reject multiple times. This is not good, but it only really happens in rare cases and where there are errors anyway.
4. Accept that the error message does not tell you what a request failed. Can be mitigated by manipulating the error message on the fly (hacky) or by adding a note into the docs ([here](https://grammy.dev/guide/deployment-types.html#ending-webhook-requests-in-time)).
5. Not merging.

Here is how much I like those options on a scale of “I'm fine with that” to “Not really an option”: 4-2-1-3-5

@dzek69 what is your opinion?

@KnightNiwrem @trgwii @wojpawlik @EdJoPaTo feel free to leave a comment here, too, if you have time.

Reply:
Well. Multiple rejections will probably never cause any real issue, won't crash Node/other engines or anything.

1. That's why 1 seem pointless if the code would stay very similar to what we currently have.
2. again is pointless. The code smell will persist, only internally it won't execute, but it does not matter.
3. I'd keep that. I personally leave not urgent stuff like that unfixed with a @TODO comment. It may get fixed when I'm bored or when better solution will become available. Sometimes it may never be fixed. Or after 10 years when it actually start causing real trouble (like JS engines starting to crash in such situation). I know people may not like having tens of @TODOs in the project but I think that's a sane compromise sometimes. I don't know what do you think about it.
4. We can as well fix only for Deno and leave Node with less meaningful messages 🤷🏻‍♂️ 
5. I don't even know what happens without that change on requests that takes more time. Can you share some info?
6. I don't know if node-fetch will allow different implementation of abort-controller on node. I had some troubles with that long time ago. Not sure if that's still an issue, but it sounds like it's worthy to give it a try. Not like I know any other implementation of it. But maybe forking it is actually a good idea.

Reply:
3. Fine by me.
4. That's hard, because then we'd have to extract the core part of the client into the platform-specific files, so we can write two clients. I'd rather not.
5. Requests may never terminate without timeouts. When using the default polling, this will make the bot hang indefinitely.
6. I think it's possible as long as it follows the specs. At least that's what `node-fetch` promises. Still, it is a lot of work.

Reply:
The problem with implementing 4 and ignoring the error message is that we then end up with #34 again.

Reply:
Ok, so at first I wanted to say to merge it but one got me thinking...

Regarding `5.` - in the code there is `// 500 seconds timeout, same as in Bot API` -- where you did found that information?

I took a look at the docs and found nothing, so I went through Telegram Bot Api code and found this:
- https://github.com/tdlib/telegram-bot-api/blob/e9587a0083c8b9e9c058ec71aec94055f1cd595d/telegram-bot-api/HttpServer.h#L69 - seems like there is 500 here, but I verified this and
- https://github.com/tdlib/td/blob/377bd18909b7a602b40e8bb074cdeb2830baef91/tdnet/td/net/HttpInboundConnection.cpp#L18 this is called `idle_timeout`. `idle` got me thinking again because I remember (but my memory sometimes are making jokes on me) having a huge upload on my personal (slow to upload) network (with self-hosted bot api). It took around 20 minutes IIRC but uploaded just fine. So I dug deeper and
- https://github.com/tdlib/td/blob/377bd18909b7a602b40e8bb074cdeb2830baef91/tdnet/td/net/HttpConnectionBase.cpp#L42 this timeout seems to be reset as long something is happening with the connection: https://github.com/tdlib/td/blob/377bd18909b7a602b40e8bb074cdeb2830baef91/tdnet/td/net/HttpConnectionBase.cpp#L141 (this last part I'm not sure I understand correctly, but it seems just a reading loop to me)

So, to sum up:
- having a timeout is nice, I'd personally want a low timeout of ~10-30 seconds for most requests, but it seems that hardcoded value of 500 seconds is just wrong and it should be configurable. Either globally but I'd more prefer to have it per request

Reply:
I don't get what #34 has common with this issue. You mean jus the "weird reaction" part from the title?

Reply:
Source of 500 seconds: https://t.me/tdlibchat/23377

If you pass a file path that doesn't exist, the stream that makes up the request body will throw, and fetch fails the request with a cryptic message. We need to catch the error from the iterator before it reaches fetch.

Reply:
I'm generally against any kind of configuration unless there's a very good reason to introduce it. For example, shorter timeouts are easily added via API transformers. If you wanted longer timeouts, that would be something to think about.

Is there a good reason why people should concern themselves with a timeout value of each API call?

Reply:
You are right - there is no need to make people concerned - so we should provide a sane default.

But for example I personally have a good reason to want this configured - because I host my bot api server on slow network and as I said - I had uploads taking 20 minutes (so 500 seconds hard limit isn't high enough for me).

Also for high-traffic bots If I'm sending just text messages or commands - I expect it to take max 10 seconds. If it takes longer there is 99% chance it's stuck and won't finish even in those 500 seconds. I'd rather want it to fail fast and let me (or auto-retry) retry the request or consider it failed and done, so I can free the queue for other users of my bot.

But when sending media this 10 seconds timeout just doesn't fit. I need it higher for requests that I expect to take longer.

Reply:
The upload speed is an interesting thought here: With 2GB as the current filelimit 500 seconds would require 4MB/s or higher upload to avoid the timeout.

Reply:
The 500 seconds do indeed assume that very large file transfers (note that a request can potentially be up to 20 GB) are performed locally. That's btw also a reasonable thing to do, but while I'd recommend you to not push everything over the network _twice_, I still want to build a library that works in all environments. That means that grammY shouldn't limit you from doing it, no matter how inefficient it is.

There's no fixed sane default in your setup. You'd need to dynamically adjust timeouts based on file size, which is a parameter that grammY cannot know. It supports streams with no known length.

What do you think about a flag that optionally disables timeouts entirely? That way, you can keep your architecture, and we still have our sane default.

Reply:
> If the selfhosted server is hosted at the same site the upload to the selfhosted Server is fast → no problem.

@EdJoPaTo actually not. API server must successfully upload the data to the Telegram servers to be able to respond with success result. So even if API server would read the data as fast as possible it would still wait with no transfer between the bot before sending a reply. But it actually reads the data from the bot not faster than it sends it out to Telegram servers to prevent keeping slow but huge uploads in the memory (I guess, but this sounds reasonable). At least that how it looks when I upload stuff on my slow network.

Reply:
> request can potentially be up to 20 GB
Isn't the limit around 5GB?

Globally disabling timeouts seems ok, but only ok. I don't want GrammY to guess the timeout - that sounds like out of the scope of a library. So I'm still all for having this configurable but to prevent having more parameters withing the api methods we could add something like `buildUrl` but for timeout. I don't fully understand API transformers yet but reading your previous comment I understand this is no good for fully controlling the timeout.

Reply:
If 1-reject-per-promise is desirable (which I'm not too bothered about either), I'd imagine that a promise-only implementation would be something like (also no longer returns a giant promise):

```ts
const waitFetch = new Promise(async (resolve, reject) => {
    try {
        const res = await fetch(url, { ...options, ...config });
        resolve(res);
    } catch (err) {
        let msg = `Network request for '${method}' failed!`;
        if (isTelegramError(err)) {
            msg += ` (${err.status}: ${err.statusText})`;
        }
        if (opts.sensitiveLogs && err instanceof Error) {
            msg += ` ${err.message}`;
        }
        reject(new HttpError(msg, err));
    }
});

let onStreamError: Function;
const waitStreamError = new Promise((_, reject) => {
    onStreamError = (err: unknown) => {
        abort();
        reject(err);
    }
});

const config = formDataRequired
    ? createFormDataPayload(p, onStreamError)
    : createJsonPayload(p);

let timeoutHandle: any;
const waitTimeoutError = new Promise((_, reject) => {
    timeoutHandle = setTimeout(() => {
        const msg = `Request to '${method}' timed out after 500 seconds`;
        abort();
        reject(new Error(msg));
    }, 500_000);
});

// Perform request
try {
    const res = await Promise.race([
        waitFetch,
        waitStreamError,
        waitTimeoutError,
    ]);
    clearTimeout(timeoutHandle);
    const resJSON = await res.json();
    return resJSON;
} catch (err) {
    clearTimeout(timeoutHandle);
    throw err;
}
```

Reply:
This implementation is cool, encapsulating the multiple rejections behind a `Promise.race` is smart. I think we should do it that way